### PR TITLE
Add support for multiple subprojects

### PR DESCRIPTION
**Pull request by v6ak**
Adds support for multiple subproject with just single application of the plugin and with just one SBT call. This prevents issues with parallel SBT calls.

Note that function testBothModes can be used for further test simplification.

### CODE REVIEW

**Comment by v6ak** [comment]
The failure on Windows is likely related to argument quoting/escaping. Maybe I should get a Windows VM and debug it there.

File: index.ts [hunk 1]
```diff
@@ -28,24 +28,68 @@ function printSbtTask(task: string, cwd?: string): Promise<string> {
...
+  return a.map((item, i) => [item, b[i]]);
+}
 
-  const fullURIPrefix = uriPrefix ? (uriPrefix + ':') : 'scalajs:';
+export default function scalaJSPlugin(options: ScalaJSPluginOptions = {}): VitePlugin {
+  const { cwd } = options;
+  const subprojects = extractSubprojects(options);
+  // This also checks for duplicates
+  const spByProjectID = mapBy(subprojects, (p) => p.projectID, "projectID")
+  const spByUriPrefix = mapBy(subprojects, (p) => p.uriPrefix, "uriPrefix")
```
Comment (sjrd) [inline index.ts:89]: These variables are unused. So instead of a `mapBy`, consider simplifying to *only* check duplicates in an array:
```typescript
function checkNoDuplicates<T>(a: Array<T>, itemName: string): void
```
then call it as
```suggestion
  checkNoDuplicates(subprojects.map((p) => p.projectID), "projectID");
  checkNoDuplicates(subprojects.map((p) => p.uriPrefix), "uriPrefix");
```

File: test/plugin.test.ts [hunk 2]
```diff
@@ -18,6 +18,23 @@ function normalizeSlashes(path: string | null): string | null {
...
+  testFunction: (d: string, func: () => Promise<void>, options: TestOptions) => void,
+  description: string,
+  f: (mode: string, suffix: string) => Promise<void>,
+  testOptions: TestOptions,
+) {
+  testFunction ||= it
+  const MODES = [["production", MODE_PRODUCTION, "opt"], ["development", MODE_DEVELOPMENT, "fastopt"]]
+  MODES.forEach( ([modeName, mode, suffix]) => {
+    testFunction(
+      description + "(" + modeName + ")",
```
Comment (sjrd) [inline test/plugin.test.ts:31]: Nit:
```suggestion
      description + " (" + modeName + ")",
```
will probably display better.

File: test/plugin.test.ts [hunk 3]
```diff
@@ -102,6 +119,77 @@ describe("scalaJSPlugin", () => {
...
+    await plugin.configResolved.call(undefined, { mode: mode });
+    await plugin.buildStart.call(fakePluginContext, {});
+
+    expect(normalizeSlashes(await plugin.resolveId.call(fakePluginContext, 'foo:main.js')))
+      .toContain('/testproject/other-project/target/scala-3.2.2/otherproject-' + suffix + '/main.js');
+    expect(normalizeSlashes(await plugin.resolveId.call(fakePluginContext, 'bar:main.js')))
+      .toContain('/testproject/target/scala-3.2.2/testproject-' + suffix + '/main.js');
+
+    expect(await plugin.resolveId.call(fakePluginContext, 'scalajs/main.js'))
+      .toBeNull();
```
Comment (sjrd) [inline test/plugin.test.ts:145]:
```suggestion
    expect(await plugin.resolveId.call(fakePluginContext, 'scalajs:main.js'))
      .toBeNull();
```
would make more sense, wouldn't it?

File: test/plugin.test.ts [hunk 4]
```diff
@@ -102,6 +119,77 @@ describe("scalaJSPlugin", () => {
...
+        },
+        {
+          projectID: null,
+          uriPrefix: "foo",
+        },
+      ]
+    });
+  });
+
+  it.fails("when both projectID and subproojects are specified", async () => {
```
Comment (sjrd) [inline test/plugin.test.ts:178]:
```suggestion
  it.fails("when both projectID and subprojects are specified", async () => {
```

File: test/plugin.test.ts [hunk 5]
```diff
@@ -102,6 +119,77 @@ describe("scalaJSPlugin", () => {
...
+    setup({
+      projectID: "xxx",
+      subprojects: []
+    });
+  });
+
+  it.fails("when both uriPrefix and subproojects are specified", async () => {
```
Comment (sjrd) [inline test/plugin.test.ts:185]:
```suggestion
  it.fails("when both uriPrefix and subprojects are specified", async () => {
```

File: test/plugin.test.ts [hunk 6]
```diff
@@ -102,6 +119,77 @@ describe("scalaJSPlugin", () => {
...
+    setup({
+      uriPrefix: "xxx",
+      subprojects: []
```
Comment (sjrd) [inline test/plugin.test.ts:188]: Is `subprojects: []` something that should be supported at all? Probably not.

File: test/plugin.test.ts [hunk 7]
```diff
@@ -102,6 +119,77 @@ describe("scalaJSPlugin", () => {
...
+    });
+  });
+
+
   it("does not work with a project that does not link", async () => {
```
Comment (sjrd) [inline test/plugin.test.ts:193]: Double blank line:
```suggestion
  });

  it("does not work with a project that does not link", async () => {
```

**Review by sjrd — CHANGES_REQUESTED** [review]
Thanks for the PR. This should be good to have, indeed.

I have some comments, but I'm mostly concerned about the comment in `printSbtTasks`.

File: index.ts [hunk 8]
```diff
@@ -28,24 +28,68 @@ function printSbtTask(task: string, cwd?: string): Promise<string> {
       if (code !== 0)
         reject(new Error(`sbt invocation for Scala.js compilation failed with exit code ${code}.`));
       else
-        resolve(fullOutput.trimEnd().split('\n').at(-1)!);
+        resolve(fullOutput.trimEnd().split('\n').slice(-tasks.length));
```
[the remark this answers is not in the document]
Comment (v6ak) [reply index.ts:31]: Maybe we can. I am not sure what to do there, though. We could filter lines starting with `[info]` (and warning etc.) and assume that this is not the actual path. (Technically, many weird characters might occur there, IIRC Linux allows everything but `'\0'` and `'/'` in a file name. I am sure newlines would probably break it, too, so it depends how much we want to be perfectionists…)

Other alternative would be calling SBT twice, but this would increate startup time, as starting two SBT instances in parallel sometimes fails. This still wouldn't resolve the issue with newlines, so maybe filtering `[info]` and similar prefixes would be a better way.

File: index.ts (hunk 1 above)
Comment (v6ak) [reply index.ts:89]: I am a bit undecided there. Sure, this can be the way. OTOH, I suspect those variables might be needed soon.

File: test/plugin.test.ts (hunk 3 above)
Comment (v6ak) [reply test/plugin.test.ts:145]: Correct. BTW, the `scalajs/main.js` is copied from other tests. Maybe we should fix it in the other places, although it is out of scope of this PR.

File: test/plugin.test.ts (hunk 6 above)
Comment (v6ak) [reply test/plugin.test.ts:188]: Yeah, it is a kind of an edge case.
* Theoretically, one might use it dynamically and no subprojects would equal to disabled plugin.
* Practically, such use case is unlikely to be common.
* It also might require the plugin to handle those edge cases. (Well, I am not sure if printSbtTasks handles empty list well…)

So, I'll disable this edge case.

File: test/plugin.test.ts [hunk 9]
```diff
@@ -18,6 +18,23 @@ function normalizeSlashes(path: string | null): string | null {
   return path === null ? null : path.replace(/\\/g, '/');
 }
 
+function testBothModes(
+  testFunction: (d: string, func: () => Promise<void>, options: TestOptions) => void,
+  description: string,
+  f: (mode: string, suffix: string) => Promise<void>,
+  testOptions: TestOptions,
+) {
+  testFunction ||= it
```
[the remark this answers is not in the document]
Comment (v6ak) [reply test/plugin.test.ts]: Removed.

File: index.ts [hunk 10]
```diff
@@ -28,24 +28,68 @@ function printSbtTask(task: string, cwd?: string): Promise<string> {
...
   uriPrefix?: string,
+  subprojects?: Array<Subproject>,
 }
 
-export default function scalaJSPlugin(options: ScalaJSPluginOptions = {}): VitePlugin {
-  const { cwd, projectID, uriPrefix } = options;
+function extractSubprojects(options: ScalaJSPluginOptions): Array<Subproject> {
+  if (options.subprojects) {
+    if (options.projectID || options.uriPrefix) {
+      throw new Error("If you specify subprojects, you cannot specify projectID / uriPrefix")
```
[the remark this answers is not in the document]
Comment (v6ak) [reply index.ts]: Added.

File: test/plugin.test.ts (hunk 2 above)
Comment (v6ak) [reply test/plugin.test.ts:30]: Added.

File: test/plugin.test.ts (hunk 4 above)
Comment (v6ak) [reply test/plugin.test.ts]: Fixed.

File: test/plugin.test.ts (hunk 5 above)
Comment (v6ak) [reply test/plugin.test.ts]: Fixed.

File: test/plugin.test.ts (hunk 7 above)
Comment (v6ak) [reply test/plugin.test.ts]: Removed.
